### PR TITLE
Universal/renderpass fbfetch

### PR DESCRIPTION
**Pull request by phi-lira**
### Purpose of this PR
 - Add framebuffer fetch macros to Common.hlsl
 - Add missing _float and _half defines in Vulkan, GLES3 and Metal api files.

---
### Release Notes

---
### Testing status
**Katana Tests**: First off we need to make sure the Katana SRP tests are green?

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in devloper mode, you have a button at end of resources that check the pathes)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

Launch Katana (Link on master here - update for your branch):
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=master&automation-tools_branch=master&unity_branch=trunk&sort=1-asc

Alternative, launch Yamato (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline

Any test projects to go with this to help reviewers?

---
### Overall Product Risks
**Technical Risk**: None, Low, Medium, High?

**Halo Effect**: None, Low, Medium, High?

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl [hunk 1]
```diff
@@ -182,6 +182,47 @@
...
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_UINT_MS(idx) cbuffer hlslcc_SubpassInput_U_##idx { uint4 hlslcc_fbinput_##idx[8]; }
+
+#define UNITY_READ_FRAMEBUFFER_INPUT(idx, v2fname) hlslcc_fbinput_##idx
+#define UNITY_READ_FRAMEBUFFER_INPUT_MS(idx, sampleIdx, v2fname) hlslcc_fbinput_##idx[sampleIdx]
+
+#else
+// Renderpass inputs: General fallback path
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_FLOAT(idx) TEXTURE2D_FLOAT(_UnityFBInput##idx); float4 _UnityFBInput##idx##_TexelSize
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_HALF(idx) TEXTURE2D_HALF(_UnityFBInput##idx); float4 _UnityFBInput##idx##_TexelSize
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_INT(idx) Texture2D<int4>(_UnityFBInput##idx); float4 _UnityFBInput##idx##_TexelSize
```
Comment (aleks01010101) [inline com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: `Texture2D<int4>` and `Texture2D<uint4>` should probably say "#error unsupported" on GLES2. Unless the compiler says this anyway :)

File: com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl [hunk 2]
```diff
@@ -182,6 +182,47 @@
...
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_UINT(idx) Texture2D<uint4>(_UnityFBInput##idx); float4 _UnityFBInput##idx##_TexelSize
+
+#define UNITY_READ_FRAMEBUFFER_INPUT(idx, v2fvertexname) _UnityFBInput##idx.Load(uint3(v2fvertexname.xy, 0))
```
Comment (aleks01010101) [inline com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: GLES2 will probably complain about this as well, but if the compiler says this in compile time, we're good.

File: com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl [hunk 3]
```diff
@@ -182,6 +182,47 @@
...
+
+// MSAA input framebuffers via tex2dms
+
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_FLOAT_MS(idx) Texture2DMS<float4> _UnityFBInput##idx; float4 _UnityFBInput##idx##_TexelSize
```
Comment (aleks01010101) [inline com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: Same goes for MS textures and GLES2

**Comment by jessebarker** [comment]
LGTM.  I'm assuming you've tested this to ensure that the right input attachment descriptors get generated from these declarations (at least on the Vulkan side)?

**Comment by sebastienlagarde** [comment]
please add changelog

File: com.unity.render-pipelines.core/ShaderLibrary/API/D3D11.hlsl [hunk 4]
```diff
@@ -132,3 +132,16 @@
 #define GATHER_GREEN_TEXTURE2D(textureName, samplerName, coord2)          textureName.GatherGreen(samplerName, coord2)
 #define GATHER_BLUE_TEXTURE2D(textureName, samplerName, coord2)           textureName.GatherBlue(samplerName, coord2)
 #define GATHER_ALPHA_TEXTURE2D(textureName, samplerName, coord2)          textureName.GatherAlpha(samplerName, coord2)
+
+// Renderpass inputs: General fallback path
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_FLOAT(idx) TEXTURE2D_FLOAT(_UnityFBInput##idx); float4 _UnityFBInput##idx##_TexelSize
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_HALF(idx) TEXTURE2D_HALF(_UnityFBInput##idx); float4 _UnityFBInput##idx##_TexelSize
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.core/ShaderLibrary/API/D3D11.hlsl]: What about other texel types? Maybe it's best to use a template parameter?

File: com.unity.render-pipelines.core/ShaderLibrary/API/D3D11.hlsl [hunk 5]
```diff
@@ -132,3 +132,16 @@
...
+
+#define UNITY_READ_FRAMEBUFFER_INPUT(idx, v2fvertexname) _UnityFBInput##idx.Load(uint3(v2fvertexname.xy, 0))
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.core/ShaderLibrary/API/D3D11.hlsl]: Probably safer to use load macros here.

File: com.unity.render-pipelines.core/ShaderLibrary/API/D3D11.hlsl [hunk 6]
```diff
@@ -132,3 +132,16 @@
...
+
+// MSAA input framebuffers via tex2dms
+
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_FLOAT_MS(idx) Texture2DMS<float4> _UnityFBInput##idx; float4 _UnityFBInput##idx##_TexelSize
+#define UNITY_DECLARE_FRAMEBUFFER_INPUT_HALF_MS(idx) Texture2DMS<float4> _UnityFBInput##idx; float4 _UnityFBInput##idx##_TexelSize
```
Comment (EvgeniiG) [inline com.unity.render-pipelines.core/ShaderLibrary/API/D3D11.hlsl]: I am a bit surprised we don't actually have a tex 2d msaa macro...

**Comment by EvgeniiG** [comment]
So, it seems to me if you use TEXTURE2D, TEXTURE2D_MSAA (doesn't exist yet), LOAD_TEXTURE2D and LOAD_TEXTURE2D_MSAA macros, you don't actually have to implement this per platform?

**Comment by ellioman** [comment]
@phi-lira Can we fix the two conflicted files and merge it? :)

**Comment by phi-lira** [comment]
This will be done differently by @jonuuukas because changes in console packages. Discussion in slack.

**Pull request closed without merging**